java/vtgate-client/src/main/java/com/youtube/vitess/vtgate/Field.java

```diff
@@ -1,27 +1,214 @@
+    VT_TIMESTAMP_FLAG(1024),      /* Field is a timestamp */
+    VT_SET_FLAG(2048),        /* field is a set */
+    VT_NO_DEFAULT_VALUE_FLAG(4096),   /* Field doesn't have default value */
+    VT_ON_UPDATE_NOW_FLAG(8192),         /* Field is set to NOW on UPDATE */
+    VT_NUM_FLAG(32768),       /* Field is num (for clients) */
+    VT_PART_KEY_FLAG(16384),      /* Intern; Part of some key */
+    VT_GROUP_FLAG(32768),     /* Intern: Group field */
+    VT_UNIQUE_FLAG(65536),        /* Intern: Used by sql_yacc */
+    VT_BINCMP_FLAG(131072);       /* Intern: Used by sql_yacc */
+    
+    public int mysqlFlag;
+    
+    Flag(int mysqlFlag) {
+      this.mysqlFlag = mysqlFlag;
+    }
+  }
+
+  /**
+   * Represents all field types supported by Vitess and their corresponding types in Java. mysqlType
+   * numbers should exactly match values defined in dist/mysql-5.1.52/include/mysql/mysql_com.h
+   *
+   */
+  enum FieldType {
+    VT_DECIMAL(0, BigDecimal.class),
+    VT_TINY(1, Integer.class),
+    VT_SHORT(2, Integer.class),
+    VT_LONG(3, Long.class),
+    VT_FLOAT(4, Float.class),
+    VT_DOUBLE(5, Double.class),
+    VT_NULL(6, null),
+    VT_TIMESTAMP(7, DateTime.class),
+    VT_LONGLONG(8, Long.class, UnsignedLong.class),
+    VT_INT24(9, Integer.class),
+    VT_DATE(10, DateTime.class),
+    VT_TIME(11, DateTime.class),
+    VT_DATETIME(12, DateTime.class),
+    VT_YEAR(13, Short.class),
+    VT_NEWDATE(14, DateTime.class),
+    VT_VARCHAR(15, byte[].class),
+    VT_BIT(16, byte[].class),
+    VT_NEWDECIMAL(246, BigDecimal.class),
+    VT_ENUM(247, String.class),
+    VT_SET(248, String.class),
+    VT_TINY_BLOB(249, byte[].class),
+    VT_MEDIUM_BLOB(250, byte[].class),
+    VT_LONG_BLOB(251, byte[].class),
+    VT_BLOB(252, byte[].class),
+    VT_VAR_STRING(253, byte[].class),
+    VT_STRING(254, byte[].class),
+    VT_GEOMETRY(255, byte[].class);
+
+    public final int mysqlType;
+    public final Class javaType;
+    public final Class unsignedJavaType;
+
+    FieldType(int mysqlType, Class javaType) {
+      this.mysqlType = mysqlType;
+      this.javaType = javaType;
+      this.unsignedJavaType = javaType;
+    }
+
+    FieldType(int mysqlType, Class javaType, Class unsignedJavaType) {
+      this.mysqlType = mysqlType;
+      this.javaType = javaType;
+      this.unsignedJavaType = unsignedJavaType;
+    }
+  }
+  
   private String name;
   private FieldType type;
-
-  public Field(String name, FieldType type) {
-    this.name = name;
-    this.type = type;
+  private int flags;
+  
+  public static Field newFieldFromMysql(String name, int mysqlTypeId, int mysqlFlags) {
+    for (FieldType ft : FieldType.values()) {
+      if (ft.mysqlType == mysqlTypeId) {
+        return new Field(name, ft, mysqlFlags);
+      }
+    }
+    
+    throw new RuntimeException("Unknown MySQL type: " + mysqlTypeId);
   }
 
-  public String getName() {
-    return name;
+  public static Field newFieldForTest(FieldType fieldType, Flag flag) {
```

Comment:
Would you be fine with reducing the visibility from public to package scope?

I removed the public constructor to reduce the surface and make it less easy to construct a Field object which for example does not have a valid FieldType. In addition to that, I like the static creator method names because they are clearer e.g. "newFieldFromMysql" suggests that this isn't just a boiler plate constructor.